### PR TITLE
Return size of block (in bytes) from web3.eth.getBlock RPC function

### LINKED ISSUE #5797: Aleth's block JSON (retrieved via web3.eth.getBlock) has different "size" value than Geth's
When one retrieves blocks from Aleth via `web3.eth.getBlock(<block number>)`, the block "size" field differs from Geth's. For example, take the Genesis block and an arbitrary block with some transactions ([65003](https://etherscan.io/block/65003)) - in both cases Aleth's block size is 0 whereas Geth's varies:

### Genesis Block
Aleth size = 0, Geth size = 540
**Aleth**
```
  author: "0x0000000000000000000000000000000000000000",
  boundary: "0x0000000040000000000000000000000000000000000000000000000000000000",
  difficulty: 17179869184,
  extraData: "0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa",
  gasLimit: 5000,
  gasUsed: 0,
  hash: "0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0x0000000000000000000000000000000000000000",
  mixHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  nonce: "0x0000000000000042",
  number: 0,
  parentHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  receiptsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  seedHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 0,
  stateRoot: "0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544",
  timestamp: 0,
  totalDifficulty: 17179869184,
  transactions: [],
  transactionsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  uncles: []
```

**Geth**
```
  difficulty: 17179869184,
  extraData: "0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa",
  gasLimit: 5000,
  gasUsed: 0,
  hash: "0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0x0000000000000000000000000000000000000000",
  mixHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  nonce: "0x0000000000000042",
  number: 0,
  parentHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  receiptsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 540,
  stateRoot: "0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544",
  timestamp: 0,
  totalDifficulty: 17179869184,
  transactions: [],
  transactionsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  uncles: []
```
### Block 65003
Aleth size = 0, Geth size = 846
**Aleth**
```
  author: "0xf927a40c8b7f6e07c5af7fa2155b4864a4112b13",
  boundary: "0x0000000000932bb9e4d20f331c0fd1eda100ae03dd36d33a9b2a6a0e9cd3915e",
  difficulty: 1912573463224,
  extraData: "",
  gasLimit: 3141592,
  gasUsed: 118496,
  hash: "0x650ec59d844817aff623efdc7a976e2f7005b17035e5ac5e7025747fbc4091ad",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0xf927a40c8b7f6e07c5af7fa2155b4864a4112b13",
  mixHash: "0xdfe2a94512f29e4363b48d0e8661a78f11d41e6fd36d9f2cd91e0c77ba0d8082",
  nonce: "0x5f7c213366edf1ef",
  number: 64999,
  parentHash: "0x23ee2f4a420e1ce4beb6d5902fafa232c0d5718146e53a31b179c86949191b85",
  receiptsRoot: "0x61b8587e6a557ecefb87f0091be5b894c197a5b43d3960093bd051589351bc32",
  seedHash: "0x510e4e770828ddbf7f7b00ab00a9f6adaf81c0dc9cc85f1f8249c256942d61d9",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 0,
  stateRoot: "0x4db358e5d7eca7ead868e622e560b2c4d4588147aee681905ad7acc17e8fd8ec",
  timestamp: 1439229736,
  totalDifficulty: 73588533123492323,
  transactions: ["0xa047af96f2342ff7bc1a061b6d83a9f00af0c7986bb24924391bf0e83de22f85", "0x49be91905241a1bac6e9cfd2568be07aa4c2c13aec5c2ce4f63f27068a5836b1", "0x9acb4b6573444ed7ed12cd8920fdcded1beadacb1872228ce09b34ccfce8e008"],
  transactionsRoot: "0xac021c3d56771233aebe36ebb0442373812b0489c3b478309339991d6dd51a71",
  uncles: []
```

**Geth**
```
  difficulty: 1912573463224,
  extraData: "0x",
  gasLimit: 3141592,
  gasUsed: 118496,
  hash: "0x650ec59d844817aff623efdc7a976e2f7005b17035e5ac5e7025747fbc4091ad",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0xf927a40c8b7f6e07c5af7fa2155b4864a4112b13",
  mixHash: "0xdfe2a94512f29e4363b48d0e8661a78f11d41e6fd36d9f2cd91e0c77ba0d8082",
  nonce: "0x5f7c213366edf1ef",
  number: 64999,
  parentHash: "0x23ee2f4a420e1ce4beb6d5902fafa232c0d5718146e53a31b179c86949191b85",
  receiptsRoot: "0x61b8587e6a557ecefb87f0091be5b894c197a5b43d3960093bd051589351bc32",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 846,
  stateRoot: "0x4db358e5d7eca7ead868e622e560b2c4d4588147aee681905ad7acc17e8fd8ec",
  timestamp: 1439229736,
  totalDifficulty: 73588533123492323,
  transactions: ["0xa047af96f2342ff7bc1a061b6d83a9f00af0c7986bb24924391bf0e83de22f85", "0x49be91905241a1bac6e9cfd2568be07aa4c2c13aec5c2ce4f63f27068a5836b1", "0x9acb4b6573444ed7ed12cd8920fdcded1beadacb1872228ce09b34ccfce8e008"],
  transactionsRoot: "0xac021c3d56771233aebe36ebb0442373812b0489c3b478309339991d6dd51a71",
  uncles: []
```

### PR DESCRIPTION
**Pull request by halfalicious**
Fix #5797, #5814

Return the size of a block in bytes to calls to `web3.eth.getBlock`. This requires extending the `BlockDetails `structure (by adding a new `blockSizeBytes `field) which in turn requires incrementing the db minor version number (since `BlockDetails ` are stored in the Extras database) which triggers an automatic database rebuild (which can take a while depending on how many blocks are in your local chain).

I've also made some minor changes to the database rebuild code to handle edge cases:
* Detect a previously prematurely terminated rebuild and log an error / throw exception if detected
* Throw an exception on rebuild failure (previously only logged an error message)
* Only update the database minor version file after a successful rebuild
* Additional logging during database rebuild

Tested changes by calling web3.eth.getBlock with the following block numbers and comparing the results against Etherscan: 0, 1, 65001 (3 txs), 101146 (2 txs), 109424 (1 tx), 115367 (7 txs)

### CODE REVIEW

**Comment by halfalicious** [comment]
Size returned for genesis block is still 0 - I suspect that the genesis block isn't actually inserted into the db

**Comment by halfalicious** [comment]
Size returned for genesis block is still 0 bytes, need to investigate.

File: libethereum/BlockDetails.cpp [hunk 1]
```diff
@@ -14,14 +14,13 @@ BlockDetails::BlockDetails(RLP const& _r)
 {
 	number = _r[0].toInt<unsigned>();
```
Comment (gumb0) [inline libethereum/BlockDetails.cpp]: Could you also convert tabs to spaces in BlockDetails.cpp and BlockDetails.h?

File: libethereum/BlockDetails.h [hunk 2]
```diff
@@ -17,27 +17,39 @@ namespace eth
...
 	u256 totalDifficulty = Invalid256;
-	h256 parent;
-	h256s children;
+	h256 parentHash;
+	h256s childHashes;
 
-	mutable unsigned size;
+    // Size of the BlockDetails RLP (in bytes). Used for computing blockchain memory usage
+    // statistics. Field name must be 'size' as BlockChain::getHashSize depends on this
+    mutable unsigned size;
```
Comment (gumb0) [inline libethereum/BlockDetails.h]: Is it still useful for anything?

File: libweb3jsonrpc/Eth.cpp [hunk 3]
```diff
@@ -345,17 +345,28 @@ Json::Value Eth::eth_getBlockByHash(string const& _blockHash, bool _includeTrans
...
 
-		if (_includeTransactions)
-			return toJson(client()->blockInfo(h), client()->blockDetails(h), client()->uncleHashes(h), client()->transactions(h), client()->sealEngine());
-		else
-			return toJson(client()->blockInfo(h), client()->blockDetails(h), client()->uncleHashes(h), client()->transactionHashes(h), client()->sealEngine());
-	}
-	catch (...)
-	{
+        Json::Value ret;
+        auto const blockDetails = client()->blockDetails(h);
```
Comment (gumb0) [inline libweb3jsonrpc/Eth.cpp]: Could you then also extract here the calls ro `blockInfo`, `uncleHashes`, `sealEngine`, not to duplicate them below

File: libweb3jsonrpc/Eth.cpp [hunk 4]
```diff
@@ -345,17 +345,28 @@ Json::Value Eth::eth_getBlockByHash(string const& _blockHash, bool _includeTrans
...
+            ret = toJson(client()->blockInfo(h), blockDetails,
+                client()->uncleHashes(h), client()->transactions(h), client()->sealEngine());
+        else
+            ret = toJson(client()->blockInfo(h), blockDetails,
+                client()->uncleHashes(h), client()->transactionHashes(h), client()->sealEngine());
+
+        // We need to explicitly set the "size" field to the block size in bytes since the
+        // BlockDetails "size" field refers to something else (size of BlockDetails RLP) and cannot
+        // be changed
+        ret["size"] = toJS(blockDetails.blockSizeBytes);
```
Comment (gumb0) [inline libweb3jsonrpc/Eth.cpp]: Why not do this inside `toJson(BlockHeader, BlockDetails, UncleHashes, Transactions, SealEngineFace*)` ?

File: libethereum/BlockChain.cpp [hunk 5]
```diff
@@ -206,18 +206,15 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
     {
         if (_we == WithExisting::Kill)
         {
-            cnote << "Killing blockchain (" << chainSubPathBlocks << ") & extras ("
-                  << extrasSubPathExtras << ") databases (WithExisting::Kill).";
-            fs::remove_all(chainSubPathBlocks);
-            fs::remove_all(extrasSubPathExtras);
+            LOG(m_loggerInfo) << "Deleting all databases. This will require a resync from genesis.";
+            fs::remove_all(chainPath);
```
Comment (gumb0) [inline libethereum/BlockChain.cpp]: I think here it was removing only blocks and extras because `BlockChain` class is responsible on only for these DB, and `State` is responsible for state.

I'd leave it without change for better separation of concertns.
(with this change there's at least a problem when it gets to `State::openDB` and logs `Killing state database` - that's not true anymore)

File: libethereum/BlockChain.cpp [hunk 6]
```diff
@@ -206,18 +206,15 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
...
         }
 
         fs::create_directories(extrasPath);
         DEV_IGNORE_EXCEPTIONS(fs::permissions(extrasPath, fs::owner_all));
 
         auto const extrasSubPathMinor = extrasPath / fs::path("minor");
```
Comment (gumb0) [inline libethereum/BlockChain.cpp]: Perhaps make a constant out of `"minor"` as it's used in more that one place now

File: libethereum/BlockChain.cpp (hunk 5 above)
Comment (gumb0) [reply libethereum/BlockChain.cpp]: Ah I see below that you did it to remove also extras.old, if it's there... Maybe better to remove existing extras.old in that case?

File: libethereum/BlockChain.cpp [hunk 7]
```diff
@@ -384,16 +398,19 @@ void BlockChain::rebuild(fs::path const& _path, std::function<void(unsigned, uns
...
     // first block.
-    auto const genesisDetails = BlockDetails{0, s.info().difficulty(), h256(), {}};
+    auto const genesisDetails =
+        BlockDetails{0 /* block number */, s.info().difficulty(), h256{} /* parent */,
+            {} /* children */, static_cast<unsigned>(m_params.genesisBlock().size())};
     m_details[m_genesisHash] = genesisDetails;
-    auto const genesisDetailsRlp = genesisDetails.rlp();
     m_extrasDB->insert(
-        toSlice(m_genesisHash, ExtraDetails), (db::Slice)dev::ref(genesisDetailsRlp));
+        toSlice(m_genesisHash, ExtraDetails), (db::Slice)dev::ref(genesisDetails.rlp()));
```
Comment (gumb0) [inline libethereum/BlockChain.cpp]: I think this is dangerous and might crash. `genesisDetails.rlp()` creates temporary `bytes`, then `dev::ref` creates `vector_ref` that stores the pointer to the data of that temporary, but temporary is not guaranteed to exist after `vector_ref(bytes const&)` is finished

File: libethereum/ClientBase.cpp [hunk 8]
```diff
@@ -417,9 +417,11 @@ BlockHeader ClientBase::pendingInfo() const
...
 BlockDetails ClientBase::pendingDetails() const
 {
-    auto pm = postSeal().info();
-    auto li = Interface::blockDetails(LatestBlock);
-    return BlockDetails((unsigned)pm.number(), li.totalDifficulty + pm.difficulty(), pm.parentHash(), h256s{});
+    auto pendingHeader = postSeal().info();
+    auto latestDetails = Interface::blockDetails(LatestBlock);
+    return BlockDetails{(unsigned)pendingHeader.number(),
+        latestDetails.totalDifficulty + pendingHeader.difficulty(), pendingHeader.parentHash(),
+        h256s{} /* children */, static_cast<unsigned>(postSeal().blockData().size())};
```
Comment (gumb0) [inline libethereum/ClientBase.cpp]: (`static_cast` will not be needed if you change the type of `blockSizeBytes` to `size_t`)

**Comment by halfalicious** [comment]
Need to also handle case of the minor version file not existing and the DB format changing

File: libethereum/ClientBase.cpp (hunk 8 above)
Comment (halfalicious) [reply libethereum/ClientBase.cpp]: @gumb0: had originally used size_t but the returned block sized were zero for some reason..will try again and root cause why.

File: libweb3jsonrpc/Eth.cpp (hunk 4 above)
Comment (halfalicious) [reply libweb3jsonrpc/Eth.cpp]: Great suggestion, will do that

File: libethereum/BlockDetails.h (hunk 2 above)
Comment (halfalicious) [reply libethereum/BlockDetails.h]: @gumb0 Not sure, will see if it can be removed

File: libethereum/BlockChain.cpp (hunk 5 above)
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: I think we should leave extras.old (unless one explicitly specifies --kill) since there's the possibility of using it to restore your extras DB (if you know what you're doing), keeping it around gives the user the chance to do that.

File: libethereum/BlockChain.cpp (hunk 6 above)
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: @gumb0 I don't see where it's being used multiple times in a function?

File: libethereum/ClientBase.cpp (hunk 8 above)
Comment (halfalicious) [reply libethereum/ClientBase.cpp]: Looks like that was a different issue that I've since addressed, size_t works

Comment (halfalicious) [reply libethereum/ClientBase.cpp]: Actually looks like that was an issue that I've since addressed, it works now

File: libethereum/BlockChain.cpp (hunk 6 above)
Comment (gumb0) [reply libethereum/BlockChain.cpp]: In another function (`rebuild`)

Comment (halfalicious) [reply libethereum/BlockChain.cpp]: Ah, yes, most of the paths are used in both places. I can move them to anonymous namespace vars.

File: libethereum/BlockDetails.h (hunk 2 above)
Comment (halfalicious) [reply libethereum/BlockDetails.h]: It's used to compute BlockChain memory usage stats which can be used to garbage collect data from the BlockChain cache:

Memory usage stats computation: https://github.com/ethereum/aleth/blob/43780e1cfaad9d887b5cd730a5df6abcea81f302/libethereum/BlockChain.cpp#L1184-L1205

Garbage collection: https://github.com/ethereum/aleth/blob/43780e1cfaad9d887b5cd730a5df6abcea81f302/libethereum/BlockChain.cpp#L1207-L1271

Comment (gumb0) [reply libethereum/BlockDetails.h]: Then I think you should still set it in ` BlockDetails::rlp()` function

Comment (halfalicious) [reply libethereum/BlockDetails.h]: @gumb0 Oops yes I definitely should 😀

**Comment by halfalicious** [comment]
Will squash some changes to clean up the commit history before merging

**Comment by halfalicious** [comment]
Rebased to fix changelog merge conflict

**Comment by halfalicious** [comment]
Also did a successful rebuild of ~2M blocks.

File: libethereum/ClientBase.cpp (hunk 8 above)
Comment (gumb0) [reply libethereum/ClientBase.cpp]: So can `static_cast` me removed?

File: libethereum/DatabasePaths.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace dev
+{
+namespace eth
+{
+namespace database_paths
+{
+bool databasePathsSet();
+void setDatabasePaths(boost::filesystem::path const& _rootPath, h256 const& _genesisHash);
+boost::filesystem::path rootPath();
```
Comment (gumb0) [inline libethereum/DatabasePaths.h]: I think I would return const references from these, to avoid copying

File: libethereum/DatabasePaths.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
+// Aleth: Ethereum C++ client, tools and libraries.
+// Copyright 2014-2019 Aleth Authors.
+// Licensed under the GNU General Public License, Version 3.
+
+#include "DatabasePaths.h"
+#include "libdevcore/CommonIO.h"
+#include "libdevcore/DBFactory.h"
+#include "libethcore/Exceptions.h"
```
Comment (gumb0) [inline libethereum/DatabasePaths.cpp]: Is this needed?

File: libethereum/DatabasePaths.h (hunk 9 above)
Comment (gumb0) [reply libethereum/DatabasePaths.h]: Somewhat confusing that we have `db::databasePath()` and now these. Should all these maybe better put into `DBFactory.h`?

Comment (halfalicious) [reply libethereum/DatabasePaths.h]: @gumb0 I considered that but DBFactory strikes me as more general purpose..it doesn’t include anything related to chains and such. Should I maybe rename the new namespace and files to something like BlockchainDatabasePaths to make the differences more clear?

File: libethereum/DatabasePaths.cpp (hunk 10 above)
Comment (halfalicious) [reply libethereum/DatabasePaths.cpp]: Ah nice catch - it used to be since I would throw if an empty genesis hash was supplied, but I pulled that out because tests use empty hashes. I’ll remove the header.

File: libethereum/DatabasePaths.h (hunk 9 above)
Comment (gumb0) [reply libethereum/DatabasePaths.h]: The namespace and files are fine, but then maybe it makes sense to move `db::setDatabasePath` to this file and/or merge with `setDatabasePaths`?

Isn't `rootPath` here basically the same as `db::databasePath` ?

Comment (halfalicious) [reply libethereum/DatabasePaths.h]: @gumb0 Yes, rootPath is the same as db::databasePath. After looking at this a little more I like your original suggestion of merging these together...I'll move the DatabasePaths functions out of that namespace and into the dev::db namespace.

**Comment by halfalicious** [comment]
Still need to merge DatabasePaths with dev::db

File: libethereum/DatabasePaths.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace dev
+{
+namespace eth
+{
+namespace database_paths
+{
+bool databasePathsSet();
+void setDatabasePaths(boost::filesystem::path const& _rootPath, h256 const& _genesisHash);
+boost::filesystem::path const& rootPath();
```
Comment (gumb0) [inline libethereum/DatabasePaths.h]: Also make them `noexcept`

File: libethereum/ClientBase.cpp [hunk 12]
```diff
@@ -417,9 +417,12 @@ BlockHeader ClientBase::pendingInfo() const
 
 BlockDetails ClientBase::pendingDetails() const
 {
-    auto pm = postSeal().info();
-    auto li = Interface::blockDetails(LatestBlock);
-    return BlockDetails((unsigned)pm.number(), li.totalDifficulty + pm.difficulty(), pm.parentHash(), h256s{});
+    auto pendingHeader = postSeal().info();
```
Comment (gumb0) [inline libethereum/ClientBase.cpp]: Can be `const`

File: libethereum/ClientBase.cpp [hunk 13]
```diff
@@ -417,9 +417,12 @@ BlockHeader ClientBase::pendingInfo() const
...
+    auto latestDetails = Interface::blockDetails(LatestBlock);
+    return BlockDetails
+    {static_cast<unsigned>(pendingHeader.number()),
```
Comment (gumb0) [inline libethereum/ClientBase.cpp]: weird indentation, is this clang-format?

File: libethereum/State.cpp [hunk 14]
```diff
@@ -42,40 +44,39 @@ State::State(State const& _s):
...
 OverlayDB State::openDB(fs::path const& _basePath, h256 const& _genesisHash, WithExisting _we)
 {
-    auto const path = _basePath.empty() ? db::databasePath() : _basePath;
-    auto const chainPath = path / fs::path(toHex(_genesisHash.ref().cropped(0, 4))) /
-                           fs::path(toString(c_databaseVersion));
-    auto const statePath = chainPath / fs::path("state");
-
     if (db::isDiskDatabase())
     {
+        db_paths::setDatabasePaths(_basePath, _genesisHash);
```
Comment (gumb0) [inline libethereum/State.cpp]: You call this from here and in `BlockChain` class, so it will be called at least twice, won't it?
Also it makes it impossible to have two instantces of `State` pointing to different databases.

It should rather be called once somewhere at the higher level (like main function or `Client` initialization).

File: libethereum/State.cpp [hunk 15]
```diff
@@ -42,40 +44,39 @@ State::State(State const& _s):
...
+        clog(VerbosityDebug, "statedb")
+            << "Ensuring permissions are set for path: " << db_paths::chainPath();
+        DEV_IGNORE_EXCEPTIONS(fs::permissions(db_paths::chainPath(), fs::owner_all));
     }
 
     try
     {
-        clog(VerbosityTrace, "statedb") << "Opening state database: " << statePath;
-        std::unique_ptr<db::DatabaseFace> db = db::DBFactory::create(statePath);
+        clog(VerbosityTrace, "statedb") << "Opening state database";
```
Comment (gumb0) [inline libethereum/State.cpp]: Why did you remove the path here?

File: libethereum/BlockChain.cpp [hunk 16]
```diff
@@ -226,61 +219,78 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
...
+        LOG(m_loggerError) << "Unknown error occurred when opening database. Exception details: "
+                           << ex.what();
         throw;
     }
 
     if (_we != WithExisting::Verify && !rebuildNeeded && !details(m_genesisHash))
     {
-        BlockHeader gb(m_params.genesisBlock());
+        bytes const genesisBlockBytes = m_params.genesisBlock();
+        BlockHeader gb{genesisBlockBytes};
```
Comment (gumb0) [inline libethereum/BlockChain.cpp]: can be `const`

File: libethereum/ClientBase.cpp (hunk 13 above)
Comment (halfalicious) [reply libethereum/ClientBase.cpp]: @gumb0 Yes, clang did that...i can manually format it a little nicer

File: libethereum/State.cpp (hunk 15 above)
Comment (halfalicious) [reply libethereum/State.cpp]: @gumb0 In case we're running memoryDB

File: libethereum/BlockChain.cpp (hunk 16 above)
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: Ah nice catch, thanks!

File: libethereum/State.cpp (hunk 14 above)
Comment (halfalicious) [reply libethereum/State.cpp]: @gumb0 For context, when would we ever have two `State `instances pointing to different databases?

File: libethereum/ClientBase.cpp (hunk 12 above)
Comment (halfalicious) [reply libethereum/ClientBase.cpp]: Nice catch, thanks!

File: libethereum/State.cpp (hunk 14 above)
Comment (gumb0) [reply libethereum/State.cpp]: Maybe in tests, maybe nowhere, but it's one thing to set the global data once and then only read it in various places, but writing global data from various places just doesn't seem good practice - both from thread-safety perspective and from general complexity management perspective

**Comment by halfalicious** [comment]
Rebased to address CHANGELOG merge conflict

**Comment by halfalicious** [comment]
Need to fix OpenDB test error

File: libethereum/BlockChain.cpp [hunk 17]
```diff
@@ -194,30 +194,26 @@ void BlockChain::init(ChainParams const& _p)
...
-    auto const chainPath = path / fs::path(toHex(m_genesisHash.ref().cropped(0, 4)));
-    auto const chainSubPathBlocks = chainPath / fs::path("blocks");
-    auto const extrasPath = chainPath / fs::path(toString(c_databaseVersion));
-    auto const extrasSubPathExtras = extrasPath / fs::path("extras");
     unsigned lastMinor = c_databaseMinorVersion;
     bool rebuildNeeded = false;
 
     if (db::isDiskDatabase())
     {
+        if (!_path.empty())
```
[the remark this answers is not in the document]
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: @gumb0: We can, but that seems unnecessary if m_dbPaths is already set?

File: libethereum/BlockChain.cpp [hunk 18]
```diff
@@ -226,88 +222,107 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
...
+    LOG(m_loggerInfo) << "Opened blockchain database. Latest block hash: " << currentHash()
+                      << (!rebuildNeeded ? "(rebuild not needed)" : "*** REBUILD NEEDED ***");
     return rebuildNeeded;
 }
 
 void BlockChain::open(fs::path const& _path, WithExisting _we, ProgressCallback const& _pc)
 {
     if (open(_path, _we) || _we == WithExisting::Verify)
-        rebuild(_path, _pc);
+        rebuild(fs::path{}, _pc);
```
[the remark this answers is not in the document]
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: @gumb0: Wanted a way to avoid unnecessarily reconstructing the DatabasePaths object...I don’t think we need to recreate it in rebuild if we already have built m_dbPaths

File: libethereum/BlockChain.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (gumb0) [reply libethereum/BlockChain.cpp]: You could then just 
```
if (!m_dbPaths.pathsSet()) 
  m_dbPaths.setPaths(_path, m_genesisHash);
```

But I would probably make it `unique_ptr` and
```
if (!m_dbPaths) 
  m_dbPaths = std::make_unique<DatabasePaths>(_path, m_genesisHash);
```

[the remark this answers is not in the document]
Comment (halfalicious) [reply libethereum/BlockChain.cpp]: @gumb0 unique_ptr makes sense, but I like the current behavior of only setting the path if the supplied path arg is empty since I think that's more intuitive, i.e. if I call a function and pass it an arg I expect that arg to be used and not silently dropped. What do you think?

File: libethereum/DatabasePaths.h [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+namespace dev
+{
+namespace eth
+{
+class DatabasePaths
+{
+public:
+    DatabasePaths(boost::filesystem::path const& _rootPath, h256 const& _genesisHash) noexcept;
+    DatabasePaths() = default;
+    void setPaths(boost::filesystem::path const& _rootPath, h256 const& _genesisHash) noexcept;
```
[the remark this answers is not in the document]
Comment (halfalicious) [reply libethereum/DatabasePaths.h]: @gumb0 I like the immutable idea, I went with the current approach since I was hesitant to use dynamic allocation for something trivial like paths. I can use unique_ptr and remove the default ctor + setter method which will effectively make the class immutable.

File: libethereum/DatabasePaths.h [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    bool pathsSet() const noexcept;
+    boost::filesystem::path const& rootPath() noexcept;
```
[the remark this answers is not in the document]
Comment (halfalicious) [reply libethereum/DatabasePaths.h]: Ah of course 😀

File: libethereum/BlockChain.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (gumb0) [reply libethereum/BlockChain.cpp]: Makes sense, but I don't like the hidden features of the function interface, like magic value "empty path" meaning "use previous path".
How about instead of empty path calling it with the current path (as before) and then here:
```
if (!m_dbPaths || m_dbPaths.rootPath() != _path) 
  m_dbPaths = std::make_unique<DatabasePaths>(_path, m_genesisHash);
```

then it's kind of an optimization to not reset the paths when not needed. And in practice re-creation will never happen, because it's called with the same path only.

File: libethereum/BlockChain.cpp [hunk 21]
```diff
@@ -226,74 +220,93 @@ bool BlockChain::open(fs::path const& _path, WithExisting _we)
...
         }
-        
-        cerror << "Unknown error occurred. Exception details: " << ex.what();
+
+        LOG(m_loggerError) << "Unknown error occurred when opening database. Exception details: "
+                           << ex.what();
         throw;
     }
 
     if (_we != WithExisting::Verify && !rebuildNeeded && !details(m_genesisHash))
```
Comment (halfalicious) [inline libethereum/BlockChain.cpp]: I can slightly streamline things by initializing rebuildNeeded to (_we == WithExisting::Verify) and then checking !rebuildNeeded here rather than also checking _we. This would also enable me to skip the minor version checks if rebuildNeeded == true. However, this file already has a ton of churn and this streamlining is pretty minor so I can take care of it in a future change.

**Comment by halfalicious** [comment]
Squashed a bunch of commits and rebased
